IRremoteESP8266.cpp

```diff
@@ -1955,48 +1971,55 @@ bool ICACHE_FLASH_ATTR IRrecv::decodeDaikin(decode_results *results) {
   return true;
 }
 
-// Denon, from https://github.com/z3t0/Arduino-IRremote/blob/master/ir_Denon.cpp
-bool ICACHE_FLASH_ATTR IRrecv::decodeDenon (decode_results *results) {
-	unsigned long data   = 0;  // Somewhere to build our code
-	int offset = 1;  // Skip the Gap reading
+// Decode a Denon message.
+//
+// Args:
+//   results: Ptr to the data to decode and where to store the decode result.
+//   nbits:   Expected nr. of data bits. (Typically DENON_BITS)
+// Returns:
+//   boolean: True if it can decode it, false if it can't.
+//
+// Ref: https://github.com/z3t0/Arduino-IRremote/blob/master/ir_Denon.cpp
+bool ICACHE_FLASH_ATTR IRrecv::decodeDenon(decode_results *results,
+                                           uint16_t nbits,
+                                           bool strict) {
+  // Check we have enough data
+  if (results->rawlen < 2 * nbits + HEADER + FOOTER)
+    return false;
+  if (strict && nbits != DENON_BITS)
+    return false;
 
-	// Check we have the right amount of data
-	if (results->rawlen != 1 + 2 + (2 * DENON_BITS) + 1) {
-	  return false;
-	}
+  uint64_t data = 0;
+  uint16_t offset = OFFSET_START;
 
-	// Check initial Mark+Space match
-	if (!matchMark (results->rawbuf[offset++], DENON_HDR_MARK )) {
-	  return false;
-	}
-	if (!matchSpace(results->rawbuf[offset++], DENON_HDR_SPACE)) {
-	  return false;
-	}
-
-	// Read the bits in
-	for (int i = 0;  i < DENON_BITS;  i++) {
-		// Each bit looks like: DENON_MARK + DENON_SPACE_1 -> 1
-		//                 or : DENON_MARK + DENON_SPACE_0 -> 0
-		if (!matchMark(results->rawbuf[offset++], DENON_BIT_MARK)) {
-		  return false;
-		}
-
-		// IR data is big-endian, so we shuffle it in from the right:
-		if (matchSpace(results->rawbuf[offset], DENON_ONE_SPACE)) {
-		  data = (data << 1) | 1;
-		} else if (matchSpace(results->rawbuf[offset], DENON_ZERO_SPACE)) {
-		  data = (data << 1) | 0;
-		} else {
-		  return false;
-		}
-		offset++;
-	}
-
-	// Success
-  results->bits = DENON_BITS;
-	results->value = data;
-	results->decode_type = DENON;
-	return true;
+  // Header
+  if (!matchMark(results->rawbuf[offset++], DENON_HDR_MARK))
+    return false;
+  if (!matchSpace(results->rawbuf[offset++], DENON_HDR_SPACE))
+    return false;
+  // Data
+  for (int i = 0; i < nbits; i++) {
+    if (!matchMark(results->rawbuf[offset++], DENON_BIT_MARK))
```

Comment:
Can we move this outside the loop? Reason being, the IR Code that passes through the checks of DENON_HDR_MARK and DENON_HDR_SPACE is very likely to have a valid DENON_BIT_MARK. So instead of checking in the loop for every iteration, can we just check it once before calling the loop? I hope I am making sense. Please feel free to correct me if I am not. :)